activerecord/lib/active_record/autosave_association.rb

```diff
@@ -402,8 +402,15 @@ def save_collection_association(reflection)
                   saved = association.insert_record(record, false)
                 elsif !reflection.nested?
                   if reflection.validate?
-                    valid = association_valid?(reflection, record, index)
-                    saved = valid ? association.insert_record(record, false) : false
+                    association_valid?(reflection, record, index)
+                    record.errors.delete(reflection.foreign_key)
+
+                    saved =
+                      if record.errors.empty?
+                        association.insert_record(record, false)
+                      else
+                        false
+                      end
                   else
                     association.insert_record(record)
```

Comment:
@kamipo should not this set `saved`?

Reply:
At least I think should not set `saved` here for stable version.

Before #32796, autosaving associated collections always don't care about whether saving collections would be succeeded or not.
The else branch keeps the original behavior when the association `validate: false` is given explicitly.